sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -611,20 +611,33 @@ RT::PiProgram ProgramManager::getBuiltPIProgram(
 
   for (RTDeviceBinaryImage::PropertyRange::ConstIterator It : ARange) {
     using namespace std::literals;
-    if ((*It)->Name != "aspects"sv)
-      continue;
-    ByteArray Aspects = DeviceBinaryProperty(*It).asByteArray();
-    // 8 because we need to skip 64-bits of size of the byte array
-    auto *AIt = reinterpret_cast<const std::uint32_t *>(&Aspects[8]);
-    auto *AEnd =
-        reinterpret_cast<const std::uint32_t *>(&Aspects[0] + Aspects.size());
-    while (AIt != AEnd) {
-      auto Aspect = static_cast<aspect>(*AIt);
-      if (!Dev->has(Aspect))
+    if ((*It)->Name == "aspects"sv) {
+      ByteArray Aspects = DeviceBinaryProperty(*It).asByteArray();
+      // 8 because we need to skip 64-bits of size of the byte array
+      Aspects.dropBytes(8);
+      while (!Aspects.empty()) {
+        auto Aspect = static_cast<aspect>(Aspects.consume<int>());
+        if (!Dev->has(Aspect))
+          throw sycl::exception(errc::kernel_not_supported,
+                                "Required aspect " + getAspectNameStr(Aspect) +
+                                    " is not supported on the device");
+      }
+    } else if ((*It)->Name == "reqd_sub_group_size"sv) {
+      auto ReqdSubGroupSize = DeviceBinaryProperty(*It).asUint32();
+      auto SupportedSubGroupSizes =
+          Device.get_info<info::device::sub_group_sizes>();
+
+      // ReqdSubGroupSize != 1 is a WA for ESIMD, no backend
+      // currently includes 1 as a valid sub-group size.
+      // TODO: remove this WA when backends support this.
+      if (ReqdSubGroupSize != 1 &&
```

Comment:
I saw that in one of previous commits you also had a check for whether the device image is a ESIMD device image - should we return it back? Just completely ignoring `1` here technically violates the spec

Reply:
Yea, we should - I removed it because I misunderstood why the workaround wasn't completely working in the first place, so I lightened the restriction to only check the `reqd-sub-group-size` for 1. Actually, I think checking if the image is an ESIMD image suffices.